### PR TITLE
JENKINS-64365: Support SonarQube 7.7 with Pull Request Analysis

### PR DESCRIPTION
**Pull request by masooh**
> Incompatibility with Sonar 7.7
--
 
> Starting with Sonarqube 7.7 the preview mode (-Dsonar.analysis.mode=preview) was removed, making it incompatible with the plugin.
 
> Sonarqube release notes say it now has "native support for short-living branches" https://www.sonarqube.org/sonarqube-7-7/, author will make an effort to integrate these features. Contributions are appreciated
 

This PR is a contribution to the problem mentioned above. Feedback to this PR is highly appreciated as this is my first contribution to a Jenkins Plugin.
 
### Chosen solution
- GUI: Introduce a switch for Pull Request/SQ 7.7 and Preview Mode
- Workflow: As JSON file is no longer present, fetch result directly from SonarQube: https://sonarqube-server.com/api/issues/search?componentKeys=myComponent&pullRequest=4711 and create a Report object like formerly based on JSON file.
- Add dependency to [Token Macro | Jenkins plugin](https://plugins.jenkins.io/token-macro/) in order to evaluate expressions like `${GERRIT_CHANGE_NUMBER}` as pullRequestKey
- Add dependency to [SonarQube Scanner | Jenkins plugin](https://plugins.jenkins.io/sonar/) in order to reuse SonarQube server config.

### Hint
As some plugins were not compatible with Jenkins 1.625 I had to upgrade some dependencies. Maybe not all were required, but probably it was a good idea anyway to stay up to date.
- Java 7 -> 8
- Jenkins parent 2.11 -> 4.4 (this also upgrades the Jenkins version which is used for testing)

### CODE REVIEW

**Comment by svogt** [comment]
Hi @masooh - Not sure if this is getting any attention, have you also logged a Jira? I haven't found one but am interested in getting this merged :) We're currently evaluating Sonar 8.5.1

**Comment by masooh** [comment]
@svogt  I tried to create an account on Jira (https://issues.jenkins.io/login.jsp), but when I try to login, I'm always getting: "Sorry, your username and password are incorrect - please try again." or "Sorry, your userid is required to answer a CAPTCHA question correctly.". Of course I double checked my password, but with no positive outcome. Due to lack of access to Jira I therefore created this PR directly.

**Comment by svogt** [comment]
> @svogt I tried to create an account on Jira (https://issues.jenkins.io/login.jsp), but when I try to login, I'm always getting: "Sorry, your username and password are incorrect - please try again." or "Sorry, your userid is required to answer a CAPTCHA question correctly.". Of course I double checked my password, but with no positive outcome. Due to lack of access to Jira I therefore created this PR directly.

Oh that's a pity, unfortunately I'm not involved with Jenkins - just stumbled upon this PR as I would like to have that too ;) 

AFAIK Jenkins uses a centralized account management system, over at https://accounts.jenkins.io/
Maybe you can sign up there and then use the login in Jira - Hopefully that should work. If that doesn't work either - write here and I'll create an issue for this.

**Comment by masooh** [comment]
@svogt Thank you for the tipp with https://accounts.jenkins.io/. This worked for me.

Created an issue now: https://issues.jenkins.io/browse/JENKINS-64365

File: src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarClient.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarClient.java:74]: 500 magic constant?

File: src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarClient.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarClient.java:90]: here it is used again should be a constant

File: src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarClient.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarClient.java:87]: why use Integer, can this be null?

File: src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarClient.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarClient.java:93]: Use com.google.common.base.Strings#isNullOrEmpty

File: src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:94]:
```suggestion
        this.serverURL = Optional.ofNullable(serverURL).map(Util::fixEmptyAndTrim).orElse(DescriptorImpl.SONAR_URL);
```

File: src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:188]: flip

File: src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:246]: sonar client must be closed

File: src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:261]: sonar client must be closed

File: src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:265]: total is an Integer, could this trigger an NPE?

File: src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:269]: Use String.format also for the other branches

File: src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Issue.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Issue.java:131]: is that correct? what about `isNew()` do we actually want this method to return null?

File: src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Issue.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Issue.java:140]: Not null-safe, same with getCreationDate

File: src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Report.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Report.java:51]: For Issue.creationDate defensive copying is used, should this also be done here?

File: src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/sonarqube/SonarConnector.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/sonarqube/SonarConnector.java:62]: SonarClient must be closed

File: src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarUtil.java
Comment (leonard84) [inline src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarUtil.java:52]: Optional.ofNullable.orElseGet

File: src/main/resources/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig/config.properties
Comment (leonard84) [inline src/main/resources/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig/config.properties:12]: Shouldn't this also mention a minimal sonar version

File: src/main/resources/org/jenkinsci/plugins/sonargerrit/config/IssueFilterConfig/config.properties
Comment (leonard84) [inline src/main/resources/org/jenkinsci/plugins/sonargerrit/config/IssueFilterConfig/config.properties:3]:
```suggestion
jenkins.plugin.settings.gerrit.filter.new.description=Reports new as well as existing issues when unchecked. For SonarQube 7 will report only new issues (Checkbox has no effect).
```

File: src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java
Comment (masooh) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:156]: com.google.common.base.Strings#isNullOrEmpty only works for Strings

File: src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/sonarqube/SonarConnector.java
Comment (masooh) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/sonarqube/SonarConnector.java:66]: Note to myself:
- use api/ce/component?component search and check that pullRequest is not in queue
- https://github.com/SonarSource/sonarqube/blob/11231d2eeced80bbc798e4f82f4c7892138675fc/sonar-ws/src/main/protobuf/ws-ce.proto#L127

File: src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java
Comment (leonard84) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:156]: right, why make this method final though?

Comment (masooh) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:156]: because it's called from the c'tor

File: src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Report.java
Comment (masooh) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Report.java:51]: Interesting, but didn't find any location in code where issues are modified apart from `issue.setNew(true)`

File: src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarUtil.java
Comment (masooh) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarUtil.java:52]: ???

File: src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Issue.java
Comment (masooh) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/entity/Issue.java:131]: it's correct (see https://stackoverflow.com/questions/39307379/java-bean-specifications-for-boolean-field), but with `is` it's more readable. -> changed

File: src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java
Comment (masooh) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig.java:265]: was never null, if so I think I will not get a result at all

File: src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/sonarqube/SonarConnector.java
Comment (masooh) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/inspection/sonarqube/SonarConnector.java]: use Duration

File: src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarUtil.java
Comment (masooh) [reply src/main/java/org/jenkinsci/plugins/sonargerrit/sonar/SonarUtil.java:52]: not the same as MoreObjects.firstNonNull

**Review by leonard84 — APPROVED** [review]

**Comment by JR-Utily** [comment]
Is this solution based on SonarQube Developer edition and its short live branches ? or does it works with a CE also ?

**Comment by leonard84** [comment]
@JR-Utily it requires the [Pull Request Analysis](https://docs.sonarqube.org/7.5/analysis/pull-request/) of the developer edition. Not really happy about it either.

**Comment by svogt** [comment]
@aquarellian - is there something left open here or can this be merged?

**Comment by leonard84** [comment]
We've been using this internally without issues for months now, and we'd be happy if we could switch back to the official plugin release once this is merged.

**Comment by IncredibleLou** [comment]
@masooh I just tested your PR on one of our test systems and I think your solution is missing the automatch functionality from the old preview mode configuration. At least for some reason our issues cannot be matched to our gerrit change.

**Comment by masooh** [comment]
@leonard84 Can you please have a look at the comment above as I have not the chance to test this any more. AFAIK I didn't change anything according to automatching, but probably I missed s.th. I thought this happens here `SonarToGerritPublisher#inspectionConfig`:

```java
//match inspection report and revision info
if (inspectionConfig.isPathCorrectionNeeded()) {
    new IssueAdapterProcessor(listener, sonarConnector, revisionInfo).process();
}
```

and this code block is unchanged.

**Comment by IncredibleLou** [comment]
@masooh  I meant you can't choose it in the UI as the checkbox to activate is in the legacy configuration block. But yes, you are right, this block is untouched, but unreachable, if you configure the pull request analysis via UI.

File: src/main/resources/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig/config.jelly
Comment (leonard84) [inline src/main/resources/org/jenkinsci/plugins/sonargerrit/config/InspectionConfig/config.jelly:63]: @IncredibleLou you mean this checkbox right?

**Comment by JR-Utily** [comment]
> @masooh I meant you can't choose it in the UI as the checkbox to activate is in the legacy configuration block. But yes, you are right, this block is untouched, but unreachable, if you configure the pull request analysis via UI.

I'm unaware of this configuration, as I'm waiting for this PR to be merged to start using the plugin. :)

But is it a blocking issue with this PR ? 
Or something that can be fixed in a later PR ?
In the first case, can you please explain what should be the correct behavior ?

**Comment by JR-Utily** [comment]
Have anyone tested this PR with https://github.com/mc1arke/sonarqube-community-branch-plugin ?

**Comment by leonard84** [comment]
> But is it a blocking issue with this PR ?
> Or something that can be fixed in a later PR ?
> In the first case, can you please explain what should be the correct behavior ?

IMHO this shouldn't block this PR and can be improved upon later.

**Comment by JR-Utily** [comment]
> > But is it a blocking issue with this PR ?
> > Or something that can be fixed in a later PR ?
> > In the first case, can you please explain what should be the correct behavior ?
> 
> IMHO this shouldn't block this PR and can be improved upon later.

@aquarellian What should be done for this PR to be merged ?

**Comment by IncredibleLou** [comment]
> > But is it a blocking issue with this PR ?
> > Or something that can be fixed in a later PR ?
> > In the first case, can you please explain what should be the correct behavior ?
> 
> IMHO this shouldn't block this PR and can be improved upon later.

It's for us just a known limitation and I can't say how many of our builds are affected, but without this PR none of our builds works, so better than nothing ;) I just hoped as the PR is a bit in limbo it could have easily be added before the merge.

**Comment by keke8273** [comment]
Hi All, thanks for the effort for creating this PR. Is there an example or documentation to show us how to use the `pull request` analysis mode in a declarative pipeline job, where I think UI configuration is not available? Thx

**Comment by keke8273** [comment]
> Have anyone tested this PR with https://github.com/mc1arke/sonarqube-community-branch-plugin ?

Yes, it does work with the `community branch plugin`. I can confirm the PR works with 

`sonarqube 8.9.2 (community edition)
  sonarqube community branch plugin 1.8.0
`

To answer my own question, the pipeline syntax can be worked out by using the `Pipeline Syntax` helper.

**Comment by thomasmay** [comment]
Please integrate this feature asap, this plugin makes no sense any more without it.

**Comment by reda-alaoui** [comment]
Thank you @masooh for opening the way !

Fixed in #110 :tada:

**Pull request closed without merging**
